src/rewards/poprewards_cache.cpp

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2019-2020 Xenios SEZC
+// https://www.veriblock.org
+// Distributed under the MIT software license, see the accompanying
+// file LICENSE or http://www.opensource.org/licenses/mit-license.php.
+
+#include <veriblock/rewards/poprewards_cache.hpp>
+
+namespace altintegration {
+
+PopRewardsBigDecimal PopRewardsCache::calculateDifficulty(
+    const index_t& /* ignore */) const {
+  auto it = buffer.crbegin();
+  // remove last element - it is the current block that we calculate difficulty
+  // for
+  ++it;
+  PopRewardsBigDecimal sumscore = 0.0;
+  size_t count = 0;
+  uint64_t difficultyInterval =
+      altParams_->getRewardParams().difficultyAveragingInterval();
+  for (; it != buffer.crend(); ++it) {
+    if (count >= difficultyInterval) {
+      break;
+    }
+    count++;
+    sumscore += it->second;
+  }
+
+  auto difficulty = sumscore / difficultyInterval;
+
+  // Minimum difficulty
+  if (difficulty < 1.0) {
+    difficulty = 1.0;
+  }
+  return difficulty;
+}
+
+std::map<std::vector<uint8_t>, int64_t> PopRewardsCache::calculatePayouts(
+    const index_t& endorsedBlock) {
+  auto popDifficulty = updateAndCalculateDifficulty(endorsedBlock);
+  auto it = std::find_if(
+      buffer.rbegin(), buffer.rend(), [&endorsedBlock](const pair_t& item) {
+        return item.first.getHash() == endorsedBlock.getHash();
+      });
+  VBK_ASSERT(it != buffer.rend() && "block should be in cache");
+  return PopRewards::calculatePayoutsInner(
+      endorsedBlock, it->second, popDifficulty);
+}
+
+PopRewardsBigDecimal PopRewardsCache::updateAndCalculateDifficulty(
+    const index_t& endorsed) {
+  cacheRebuild(endorsed);
+  return calculateDifficulty(endorsed);
+}
+
+void PopRewardsCache::cacheAddBlock(const index_t& block) {
+  auto score = scoreFromEndorsements(block);
+  buffer.push_back({block, score});
+}
+
+bool PopRewardsCache::cacheBlockExists(const index_t* block) {
+  if (block == nullptr) {
+    return false;
+  }
+  auto it =
+      std::find_if(buffer.rbegin(), buffer.rend(), [block](const pair_t& item) {
```

Comment:
since the ring buffer stores a chain, it is possible to look up blocks efficiently by height: the cache item we're looking for should be buffer.back().height - block.height elements away from the buffer end